cloud/smartos/vmadm

```diff
@@ -0,0 +1,229 @@
+
+    # vmadm create -f $manifestFile
+    if state == 'present':
+
+        # extra parameter sanity checks (we ignore uuid and alias, as they do nothing here.)
+        check_subparam(module, 'manifest', p, state)
+
+        # abort if we can't find the vm manifest file.
+        if not os.path.isfile(manifest):
+            module.fail_json(msg="could not locate manifest file: %s" % manifest)
+
+        # we need to delve into the manifest and find the alias for the VM, and then
+        # make sure that there is not a VM with that alias (even though SmartOS doesn't
+        # care about duplicated aliases).
+        # Otherwise, a new VM is created each and every single time (not idempotent).
+        data = open_and_load_json_data(module, manifest)
+
+        try:
+            vm_alias = data["alias"]
+        except:
+            module.fail_json(msg="manifest file must provide an alias for the vm")
+
+        if not vmadm_alias_lookup(module, vmadm_path, vm_alias):
+            rc, so, se = module.run_command('%s create -f %s' % (vmadm_path, manifest))
+            changed = True
+        else:
+            rc, so, se = (0, "", "")
+
+    else:
+
+        # -- common checks/tasks for all states except 'present'
+        if not uuid and not alias:
+            module.fail_json(msg="one of 'alias' or 'uuid' required when state is not 'present'")
+
+        if alias:
+            uuid = vmadm_alias_lookup(module, vmadm_path, alias)
+            if not uuid:
+                module.fail_json(msg="could not locate a matching uuid for vm alias '%s'" % alias)
+        # --
+
+        # handles most other states.
+        if state in [ 'started', 'stopped', 'rebooted', 'absent' ]:
+            rc, so, se = module.run_command('%s %s %s' % (vmadm_path, state.strip('ed').replace('opp', 'op'), uuid))
+
+            # detect whether the state was changed.
+            # NOTE: oddly enough, it uses stderr here (2014/Dec/25)
+            if se.find("Successfully") == 0:
+                changed = True
+
+        elif state == 'updated':
+            if not proplist and not manifest:
+                module.fail_json(msg="one of 'manifest' or 'proplist' required when state='updated'")
+
+            # In order to be idempotent, we need to compare the values that we're being asked to update
+            # against the values stored in the hypervisor.
+            import hashlib
+            vm_json_hash = hashlib.sha1(vmadm_get_vm_json_description(module, vmadm_path, uuid)).hexdigest()
+
+            if proplist:
+                rc, so, se = module.run_command('%s update %s %s' % (vmadm_path, uuid, proplist))
+
+            else: # manifest
+                rc, so, se = module.run_command('%s update %s -f %s' % (vmadm_path, uuid, manifest))
+
+            vm_json_newhash = hashlib.sha1(vmadm_get_vm_json_description(module, vmadm_path, uuid)).hexdigest()
```

Comment:
Checking if the vmadm update passed this way is good, but keep in might that with high amount of snapshots (reported as #282 in smartos-live repository). More performant, but complicated could be to check files in VM zone root and /etc/zones and comparing them, e.g /zones/1158cb46-37f2-4340-b060-46a36468dcd6/config/* and /etc/zones/1158cb46-37f2-4340-b060-46a36468dcd6.xml.